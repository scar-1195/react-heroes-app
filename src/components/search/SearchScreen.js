import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';

import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroByName';
import HeroCard from '../heroes/HeroCard';
import { useMemo } from 'react';

const SearchScreen = () => {

  //? useLocation es un hook que nos retorna los queries params de la ruta en la que estamos
  const location = useLocation();
  const navigate = useNavigate();

  const { q = '' } = queryString.parse( location.search );

  const [ formValues, handleInputChange ] = useForm({
    searchText: q,
  });

  const { searchText } = formValues;

  const heroesFiltered = useMemo( () => getHeroesByName(q), [q] );
  
  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${ searchText }`)
  }
  
  return (
    <>
      <h1>Super Hero Searchs</h1>
      <hr />

      <div className='row'>
        <div className='col-5'>
          <h4>Search</h4>
          <hr />

          <form onSubmit={ handleSearch }>
            <input
              autoComplete='off'
              className='form-control'
              name='searchText'
              onChange={ handleInputChange }
              placeholder='Find a super hero'
              type='text'
              value={ searchText }
            />

            <button className='btn btn-outline-primary mt-1 w-100' type='submit'>
              Find
            </button>
          </form>
        </div>

        <div className='col-7'>
          <h4>Results</h4>
          <hr />

          {
            (q === '')
              ? <div className='alert alert-info'>Find a super hero</div>
              : ( heroesFiltered.length === 0 )
                && <div className='alert alert-danger'>{q} dont exist</div>
          }

          {
            heroesFiltered.map(hero => (
              <HeroCard
                key={ hero.id }
                {... hero }
              />
            ))
          }
        </div>
      </div>
    </>
  );
}

export default SearchScreen;