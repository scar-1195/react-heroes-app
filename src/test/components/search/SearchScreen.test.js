import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import SearchScreen from '../../../components/search/SearchScreen';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate,
}));

describe('Pruebas en <SearchScreen />', () => {

  test('debe de mostrar el componente correctamente con valores por defecto', () => {

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find('.alert-info').text().trim() ).toBe( 'Find a super hero' );

  });

  test('debe de mostrara a Batman y el input con el valor del querySring', () => {

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect( wrapper.find('input').prop('value') ).toBe( 'batman' );
    expect( wrapper ).toMatchSnapshot();

  });

  test('debe de mostrar un error si no se encuentra el heroe', () => {

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchScreen />
      </MemoryRouter>
    );

    expect( wrapper.find('.alert-danger').text().trim() ).toBe( 'batman123 dont exist' );

  });

  test('debe de llamar el navigate a la nueva ruta', () => {

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <SearchScreen />
      </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
      target:{
        name: 'searchText',
        value: 'batman',
      }
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault: () => {}
    });

    expect( mockNavigate ).toHaveBeenCalledWith('?q=batman');

  });

});