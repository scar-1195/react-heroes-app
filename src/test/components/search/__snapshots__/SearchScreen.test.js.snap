// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar el componente correctamente con valores por defecto 1`] = `
Array [
  <h1>
    Super Hero Searchs
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find a super hero"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1 w-100"
          type="submit"
        >
          Find
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Find a super hero
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en <SearchScreen /> debe de mostrara a Batman y el input con el valor del querySring 1`] = `
Array [
  <h1>
    Super Hero Searchs
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find a super hero"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-1 w-100"
          type="submit"
        >
          Find
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  More info
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
